### ISSUE TITLE
Microsoft Azure Integration

### ISSUE THREAD

**Issue by arthybck** [OrigenStudio/vulcan-files#16]
Hi,

I would like to contribute to the package vulcan-files by adding the possibility to work with Microsoft Azure.
Since i'm going to work on it anyway, it could be a good option for your users.

I planned to use the "azure-storage" node package which has been well structured: https://www.npmjs.com/package/azure-storage
https://github.com/Azure/azure-storage-node

I also have some questions about your pkg: 
- Where are you, talking of the realization of your package?
- Have you planned to include other cloud providers?
- How could we process to integrate the package I am working on to yours?

I look forward to hearing from you,

Arthur

**Comment by ochicf** [comment]
Hi Arthur!

Thanks for contacting us, so glad to hear you want to contribute ^^

>Where are you, talking of the realization of your package?

Right now the package is in alpha state. We are currently using it in some projects and we keep improving it as we need new features.

We recently bumped the minor version, as we added some breaking changes. So we have: 

- `0.0.6`: initial version, less opinionated but with lesser features, and complex to use
- `0.1.0.rc-1`: more opinionated, so it can add more features out of the box so it's much easier to use (I don't think this is reflected on the documentation)

Of course everything is up to changes and there are tons of missing features we would like.

>Have you planned to include other cloud providers?

Not ourselves per se, but always had in mind that we wanted this package to be extensible with other providers. Right now I don't think it's possible without doing some changes, but they shouldn't be that hard.

My idea is that you could create a FSCollection by specifying a 3rd party provider that handles these 3 points:

- upload
- remove
- serve

I think that way it would ease to implement 3rd party providers and also to use them by the consumers.

>How could we process to integrate the package I am working on to yours?

I've been wanting to externalize the S3 features into a separate package for quite a while, so I could do that and it would serve as an example of what a 3rd party package should export.

----------------------

In conclusion, in order for you to work I think the next steps would be:

1. allow for third parties to be fully injected
2. externalize s3 as a package to be used as an example
3. update docs with 0.1.0 features (not mandatory, but would help you a lot I think)

I think I'll be able to take some time to dive into this on friday or monday.

Thanks again!

Oscar

**Comment by arthybck** [comment]
Hi Oscar,

Thank you for your fast answer.

I just started my package last week.
I suggest that we keep in touch, i will let you know when mine is ready.

Thank you for the time you will spend on it !
Have a good day.

Arthur

**Comment by ochicf** [comment]
Hi @arthybck,

Finally I've managed to make the changes I promised (sorry for the delay). I've added some documentation in the [develop](https://github.com/OrigenStudio/vulcan-files/tree/develop#24-uploading-to-3rd-parties) branch of the package.

You can also take a look of the [S3](https://github.com/OrigenStudio/vulcan-files-s3) package as an example.

I've also published the packages on atmosphere, all with versions `1.0.0-alpha.1` (used this instead of `0.1.0-rc.1` to prevent it from being updated on existing installations).

Also note that the [vulcan-files-simple-example@develop](https://github.com/OrigenStudio/vulcan-files-simple-example/tree/develop) is updated and working, so maybe you could clone it and test your integration there.

Let me know if you need anything.
